File: src/magma/app/compile/State.java
package magma.app.compile;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Stream;

public class State {
    private final List<String> rootMembers;
    private final StringBuilder buffer;

    public State(List<String> rootMembers, StringBuilder buffer) {
        this.rootMembers = rootMembers;
        this.buffer = buffer;
    }

    public State() {
        this(Collections.emptyList(), new StringBuilder());
    }

    State advance() {
        var copy = new ArrayList<>(rootMembers);
        copy.add(buffer.toString());
        return new State(copy, new StringBuilder());
    }

    public Stream<String> stream() {
        return rootMembers.stream();
    }

    public State append(char c) {
        return new State(rootMembers, buffer.append(c));
    }
}

File: src/node/magma/State.ts
/*package magma;*//*

import java.util.Collection;*//*
import java.util.stream.Stream;*//*

public class State {
	public final Collection<String> segments;*//*
	private StringBuilder buffer;*//*

	public State(Collection<String> segments, StringBuilder buffer) {
		this.segments = segments;*//*
		this.buffer = buffer;*//*
	}

	Stream<String> stream() {
		return segments.stream();*//*
	}

	State advance() {
		segments.add(buffer.toString());*//*
		this.buffer = new StringBuilder();*//*
		return this;*//*
	}

	State append(char c) {
		buffer.append(c);*//*
		return this;*//*
	}
}
*/